File: sdk/bloom.filter.ts
import murmurHash3 from "murmurhash3js";
import { decode } from "base64-arraybuffer";

// partial bloom filter implementation (read-only) that matches Go implementation at https://github.com/bits-and-blooms/bloom
export class BloomFilter {
  public bitSet: BitSet | undefined;

  constructor(
    public readonly m: number,
    public readonly k: number,
    public readonly base64Data: string
  ) {
    // create bitset lazily
    this.bitSet = undefined;
  }

  // returns true if key is in bloom filter, else false
  public has(key: string): boolean {
    if (this.bitSet == undefined) {
      this.bitSet = new BitSet(this.m, this.base64Data);
    }
    var indices = this.getIndices(key);
    for (var i = 0; i < indices.length; i++) {
      if (!this.bitSet.has(indices[i])) {
        return false;
      }
    }
    return true;
  }

  private getIndices(key: string) {
    const indices: number[] = [];
    const baseHashes = this.getBaseHashes(key);

    for (let i = 0; i < this.k; i++) {
      let ii = BigInt(i);
      let a = baseHashes[Number(ii % BigInt(2))];
      let b = (ii + (ii % BigInt(2))) % BigInt(4);
      let c = BigInt(2) + b / BigInt(2);
      let d = BigInt.asUintN(64, ii * baseHashes[Number(c)]);
      // console.log(a, b, c, d);
      const location = BigInt.asUintN(64, a + d);
      // console.log(i, location);
      // console.log(i, location % BigInt(this.m));
      indices.push(Number(location % BigInt(this.m)));
    }
    return indices;
  }

  // returns 4 64-bit hashes using murmur3 hash (x64, 128-bit implementation)
  private getBaseHashes(key: string) {
    const keyBuffer = Buffer.from(key);
    // compute 128-bit hash from key and split to get 2 64-bit hashes
    const hash1 = murmurHash3.x64.hash128(key);
    const val1 = Buffer.from(hash1.substring(0, 16), "hex").readBigUInt64BE();
    const val2 = Buffer.from(hash1.substring(16), "hex").readBigUInt64BE();
    // compute second 128-bit hash by appending 1 to key
    const hash2 = murmurHash3.x64.hash128(
      Buffer.concat([keyBuffer, Buffer.from([1])]).toString()
    );
    const val3 = Buffer.from(hash2.substring(0, 16), "hex").readBigUInt64BE();
    const val4 = Buffer.from(hash2.substring(16), "hex").readBigUInt64BE();
    return [val1, val2, val3, val4];
  }
}

class BitSet {
  private readonly data: BigUint64Array;

  constructor(public readonly m: number, public readonly base64Data: string) {
    // first Uint64 encodes m, next Uint64 encodes k, next Uint64 encodes m again (so we slice them out)
    const buffer = decode(base64Data).slice(8 * 3);
    const dataView = new DataView(buffer);
    const arrayLength = Math.ceil(this.m / 64); // how many Uint64 do we need to store m bits
    this.data = new BigUint64Array(arrayLength);
    for (let i = 0; i < arrayLength; i++) {
      this.data[i] = dataView.getBigUint64(8 * i, false); // read as big endian
    }
  }

  public has(index: number): boolean {
    const wordIndex = index >> 6;
    var wordsIndexI = index & (64 - 1);
    var mask = BigInt(Math.pow(2, wordsIndexI));
    var word = this.data[wordIndex];
    // console.log(wordIndex, wordsIndexI, mask, word)
    return (word & mask) !== BigInt(0);
  }
}

File: python-sdk/src/forta_agent/bloom_filter.py
import mmh3
import base64
import math


class BloomFilter:
    def __init__(self, dict):
        self.k = int(dict.get('k'), 16) if type(
            dict.get('k')) == str else dict.get('k')
        self.m = int(dict.get('m'), 16) if type(
            dict.get('m')) == str else dict.get('m')
        self.base64_data = dict.get('bitset')
        self.bitset = None

    def has(self, key):
        if self.bitset is None:
            self.bitset = BitSet({'m': self.m, 'bitset': self.base64_data})
        indices = self.get_indices(key)
        for index in indices:
            if not self.bitset.has(index):
                return False
        return True

    def get_indices(self, key):
        indices = []
        base_hashes = self.get_base_hashes(key)
        for i in range(self.k):
            a = base_hashes[i % 2]
            b = (i + (i % 2)) % 4
            c = 2 + (b/2)
            d = 0xFFFFFFFFFFFFFFFF & (
                i * base_hashes[int(c)])  # truncate 64 bits
            location = 0xFFFFFFFFFFFFFFFF & (a + d)  # truncate 64 bits
            indices.append(location % self.m)
        return indices

    def get_base_hashes(self, key):
        val1, val2 = mmh3.hash64(key, 0, signed=False)
        # convert key string to bytearray
        b = bytearray()
        b.extend(key.encode())
        # append 1
        b.append(1)
        val3, val4 = mmh3.hash64(bytes(b), 0, signed=False)
        return [val1, val2, val3, val4]


class BitSet:
    def __init__(self, dict):
        # first Uint64 (i.e. 8 bytes) encodes m, next Uint64 encodes k, next Uint64 encodes m again (so we slice them out)
        decoded_bytes = base64.b64decode(dict.get('bitset'))[8*3:]
        # how many Uint64 do we need to store m bits
        array_length = math.ceil(dict.get('m') / 64)
        self.data = []
        for i in range(array_length):
            bytes_index = 8*i
            self.data.append(int.from_bytes(
                decoded_bytes[bytes_index:bytes_index+8], "big"))  # read as big endian

    def has(self, index):
        word_index = index >> 6
        words_index_i = index & (64-1)
        mask = int(math.pow(2, words_index_i))
        word = self.data[word_index]
        return (word & mask) != 0
